waiter/src/waiter/mesos/marathon.clj

```diff
@@ -42,6 +42,15 @@
                             :request-method :delete
                             :spnego-auth spnego-auth))
 
+(defn delete-deployment
+  "Cancel the deployment with deployment-id.
+   No rollback deployment is created to revert the changes of deployment."
+  [{:keys [http-client marathon-url spnego-auth]} deployment-id]
+  (http-utils/http-request http-client (str marathon-url "/v2/deployments/" deployment-id)
+                           :query-string {"force" true}
```

Comment:
Do we expect passing `force` here to be better than forgoing doing anything with deployments and instead passing `force` on the `update-app` call?  If so, why is that the case?

Reply:
We want to control that an arbitrary instance does not get killed due to our update. If we did the force on the `update-app` its possible there is a race between the number of instances we knew about (the actual value can be higher during scale-up) and the number of instances we set the new value to during the update call. As a result, Marathon would then decide to arbitrarily kill an instance. We could have learnt about this instance from the scheduler syncer and sent it a request while Marathon chose to kill the instance.

Reply:
When manually killing a deployment after a change in the number of instances (scale up operation), I observed that the `instances` field was unaffected.  Given that observation, is it possible to get in a bad state, whether we're scaling up or scaling down?  Have we observed the behavior of force deleting a deployment in both cases?  How does our current code + this PR handle those cases gracefully?

Reply:
I reasoned the following scenario is possible if we do not explicitly force delete the outstanding deployment:
1. Scale-Up: Waiter asks a service to scale up to 10 instances.
2. Scale-Down: Waiter later decides to scale the service to 5 instances.
3. Scale-Down: We query marathon and find that the previous deployment is still in effect.
4. Scale-Down: We query the service state and hear back that it has 4 instances.
5. Marathon manages to scale the service to 6 instances after our previous query.
6. Syncer: Syncer hears about these 6 instances and assigns them to the router
7. Request: New client requests are assigned to these instances
8. Scale-Down: We decide to change the number of instances to 5 (we had last seen this value to be 4 in step-4, the data race we should try and protect against) using force mode
9. Marathon in actuality has already scaled the service to 6 instances so decides to arbitrarily kill an instance (that instance is potentially processing a request and hence causes that request to fail).

Do you agree that this scenario can happen?

This PR avoids the data race mentioned above by first force deleting the deployment to freeze the instances being created for the service. Hence when we subsequently query the number of instances (step-4) and when we set the new number of instances (step-8), we do not expect the number to change and avoid the data race. We would update the deployment to 6 instances instead of 5 instances since we take a max.

Reply:
We spoke offline, the issue @sradack is raising is what happens if the call to `update-app` in `scale-app` fails after we have already deleted the current deployment. We verified on Marathon that deleting an existing deployment does not affect the `task-count` and causes the service to freeze in a state where the task counts (requested instances) does not match with scheduled instance count.

An approach to tackle this scenario (while still avoiding the data race issue I mentioned above) is:
1. Trigger deployment update on every scaling cycle if the requested and scheduled instance counts do not match up. This request will be ignored if a deployment is already in flight.
2. Run scaling for a service less frequently (than the current 1-second interval). E.g. if a service has been scaled this iteration, wait at least 60 seconds before attempting to scale again. 

In effect, we want to allow time for a previous deployment to complete and avoid making superfluous deployment update calls. The duplicate calls are necessary since we do not know if we are in the frozen state and they allow us to get out of the frozen state.

@sradack what do you think of this approach?

Reply:
Since this seems localized to Marathon, it should be a Marathon-specific solution.  What do you think about setting up a loop in our Marathon code instead of inside the scaling logic?  Marathon supports embedding deployment information in the `get-app` response.  We could use that to search for the candidate apps that need to be fixed.

Reply:
Sounds good.

Reply:
Added support for `sync-deployment-maintainer`

Reply:
A race is still possible in setting the instance value with the call to `scale-app` from the auto-scaler and the sync-deployment go-blocks (it is expected to be rare). We can go the extra step and avoid the race altogether by introducing a service-level (implemented by using a hash to index into a bucket of locks) lock for scale read/write calls. Let me know what you think?

Reply:
Given the rarity of the case and the eventuality that we'll end up in a working state, I'm not sure implementing locking is worth the extra complexity.